src/main/scala/org/mauritania/main4ino/security/Auther.scala

```diff
@@ -76,31 +95,55 @@ object Auther {
-    val credentials = userCredentialsFromRequest(encry, headers, uri)
+    val credentials = userCredentialsFromRequest(headers, uri)
     val session = sessionFromRequest(headers, uri)
     for {
       user <- authenticatedUserFromSessionOrCredentials(encry, usersBy, session, credentials)
-      authorized <- checkAccess(user, method, resource)
-    } yield authorized
+    } yield user.flatMap(u => checkAccess(u, method, resource))
   }
 
   /**
-    * Create a session for a given user given a private key and a timestamp
+    * Create a session for a given user given a private key
     *
     * @param user user for whom we want to create a session
     * @param privateKey private key used for encryption of the session id to be generated
-    * @param time time used to generate the session id
     * @return a user session id
     */
-  def sessionFromUser(user: User, privateKey: CryptoBits, time: Clock): UserSession =
-    privateKey.signToken(user.id, time.millis.toString())
+  def sessionFromUser[F[_]: Sync : Monad](user: User, privateKey: CryptoBits): F[UserSession] = {
+    val claims = JWTClaims(subject = Some(user.id))
+    for {
+      key             <- HMACSHA256.buildKey[F](privateKey)
+      stringjwt       <- JWTMac.buildToString[F, HMACSHA256](claims, key)
+    } yield stringjwt
+  }
 
-  def authenticatedUserFromSessionOrCredentials(encry: EncryptionConfig, usersBy: UsersBy, session: Option[UserSession], creds: Option[(UserId, UserHashedPass)]): AuthenticationAttempt = {
-    session.flatMap(v => encry.pkey.validateSignedToken(v)).flatMap(usersBy.byId.get)
-      .orElse(creds.flatMap(usersBy.byIdPass.get))
-      .toRight(s"Could not authenticate user (user:${creds.map(_._1)} / session:${session.map(_.slice(0, 5))}...)")
+  def userIdFromSession[F[_] : Sync : Monad](session: UserSession, privateKey: CryptoBits): F[Option[UserId]] = {
+    for {
+      key             <- HMACSHA256.buildKey[F](privateKey)
+      parsed          <- JWTMac.verifyAndParse[F, HMACSHA256](session, key)
+    } yield parsed.body.subject
+  }
+
+  def userFromSession[F[_]: Sync](session: UserSession, privateKey: CryptoBits, usersBy: UsersBy): F[Option[User]] = {
+    for {
+      id <- userIdFromSession(session, privateKey)
+    } yield id.flatMap(usersBy.byId.get)
+
+  }
+
+  def authenticatedUserFromSessionOrCredentials[F[_]: Sync](encry: EncryptionConfig, usersBy: UsersBy, session: Option[UserSession], creds: Option[(UserId, UserPassword)])(implicit H: PasswordHasher[F, BCrypt]): F[AuthenticationAttempt] = {
+    for {
+      authenticatedUsrSession <- session.map(s => userFromSession(s, encry.pkey, usersBy)).sequence.map(_.flatten)
+      authenticatedUsrCreds <- creds.map { case (id, clearPass) =>
+        val configUser = usersBy.byId.get(id)
+        configUser.map { u =>
+          H.checkpw(clearPass, u.hashedpass).map(c => if (c == Verified) Some(u) else None)
+        }.sequence.map(_.flatten)
+      }.sequence.map(_.flatten)
```

Comment:
`x.map(f).sequence === x.traverse(f)`

Also, too much `flatten`? There has to be a better way.

Reply:
Maybe `OptionT` could be a way.